aspnet/migration/http-modules.rst

```diff
@@ -1,8 +1,576 @@
+
+Loading middleware options through direct injection
+---------------------------------------------------
+The options pattern has the advantage that it creates loose coupling between options values and their consumers.
+Once you've associated an options class with the actual options values, any other class can get access to the options
+through the dependency injection framework. So there is no need to pass around options values.
+
+This breaks down though if you want to use the same middleware twice, with different options. For example an authorization middleware
+used in different branches allowing different roles. You can't associate two different options objects with the one options class.
+
+The solution is to get the options objects with the actual options values in your ``Startup`` class and pass those directly to each instance of your middleware.
+
+1. Add a second key to appsettings.json
+
+    To add a second set of options to the 
+    *appsettings.json* file, simply use a new key to uniquely identify it:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/appsettings.json
+        :language: json
+        :linenos:
+        :emphasize-lines: 2-5
+
+2. Retrieve options values
+
+    In the ``Configure`` method in your ``Startup`` class, the ``Get`` method on the ``Configuration`` property lets you retrieve options values, like so:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Startup.cs
+        :language: c#
+        :linenos:
+        :lines: 1-2, 12-15, 46-48, 54-55, 62-66, 67, 70, 71, 105, 109, 110
+        :emphasize-lines: 12-16
+ 
+3. Pass options values to middleware
+
+    The ``Use...`` extension method with which you add your middleware to the pipeline would be the logical place
+    to pass in the option values:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Startup.cs
+        :language: c#
+        :linenos:
+        :lines: 1-2, 12-15, 46-48, 54-55, 62-72, 105, 109, 110
+        :emphasize-lines: 18-22
+
+4. Enable middleware to take options parameter
+
+    To make this work, you'll need an overload of your ``Use...`` extension method that takes the options parameter
+    and passes it to ``UseMiddleware``. When you call  
+    ``UseMiddleware`` with parameters, it passes those to your middleware constructor when it instantiates the middleware object.
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Middleware/MyMiddlewareWithParams.cs
+        :language: c#
+        :linenos:
+        :lines: 1-7, 48-64
+        :emphasize-lines: 17-22
+
+    Note how this wraps the options object in an ``OptionsWrapper`` object. This implements ``IOptions``, as expected by 
+    the middleware constructor:
+
+    .. literalinclude:: http-modules/sample/Asp.Net5/src/Asp.Net5/Middleware/MyMiddlewareWithParams.cs
+        :language: c#
+        :linenos:
+        :lines: 14-23
+
+
+Migrating to the new HttpContext
```

Comment:
@danroth27 would this part make sense as a stand-alone article? It seems relevant beyond the discussion of modules and handlers.